src/lightning/app/utilities/network.py

```diff
@@ -27,17 +27,56 @@
 from requests.exceptions import ConnectionError, ConnectTimeout, ReadTimeout
 from urllib3.util.retry import Retry
 
+from lightning.app.core import constants
 from lightning.app.utilities.app_helpers import Logger
 
 logger = Logger(__name__)
 
 
+# Global record to track ports that have been allocated in this session.
+_reserved_ports = set()
+
+
 def find_free_network_port() -> int:
     """Finds a free port on localhost."""
-    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-    s.bind(("", 0))
-    port = s.getsockname()[1]
-    s.close()
+    if constants.LIGHTNING_CLOUDSPACE_HOST is not None:
+        # If in a cloudspace, look for a port in the exposed range
+        for port in range(
+            constants.APP_SERVER_PORT,
+            constants.APP_SERVER_PORT + constants.LIGHTNING_CLOUDSPACE_EXPOSED_PORT_COUNT,
+        ):
+            if port in _reserved_ports:
+                continue
+
+            try:
+                s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+                s.bind(("", port))
+                s.close()
+                _reserved_ports.add(port)
+                return port
+            except OSError:
+                continue
+
+        # This error should never happen. An app using this many ports would probably fail on a single machine anyway.
+        raise RuntimeError(f"All {constants.LIGHTNING_CLOUDSPACE_EXPOSED_PORT_COUNT} ports are already in use.")
+
+    port = None
+    i = 0
+
+    while i < 10 and (port is None or port in _reserved_ports):
```

Comment:
I would rather make it a `for` loop to be sure it may not be accidently infinite

Reply:
Done!